fedimint-client/src/db.rs

```diff
@@ -262,3 +276,221 @@ impl_db_lookup!(
     key = ClientModuleRecovery,
     query_prefix = ClientModuleRecoveryPrefix
 );
+
+/// `ClientMigrationFn` is a function that modules can implement to "migrate"
+/// the database to the next database version.
+pub type ClientMigrationFn = for<'r, 'tx> fn(
+    &'r mut DatabaseTransaction<'tx>,
+    ModuleInstanceId,
+    Vec<DynState>, // active states
+    Vec<DynState>, // inactive states
+) -> Pin<
+    Box<
+        dyn futures::Future<Output = anyhow::Result<Option<(Vec<DynState>, Vec<DynState>)>>>
+            + Send
+            + 'r,
+    >,
+>;
+
+/// `apply_migrations_client` iterates from the on disk database version for the
+/// client module up to `target_db_version` and executes all of the migrations
+/// that exist in the migrations map, including state machine migrations.
+/// Each migration in the migrations map updates the database to have the
+/// correct on-disk data structures that the code is expecting. The entire
+/// process is atomic, (i.e migration from 0->1 and 1->2 happen atomically).
+/// This function is called before the module is initialized and as long as the
+/// correct migrations are supplied in the migrations map, the module
+/// will be able to read and write from the database successfully.
+pub async fn apply_migrations_client(
+    db: &Database,
+    kind: String,
+    target_db_version: DatabaseVersion,
+    migrations: BTreeMap<DatabaseVersion, ClientMigrationFn>,
+    module_instance_id: ModuleInstanceId,
+) -> Result<(), anyhow::Error> {
+    db.ensure_global()?;
+
+    let mut global_dbtx = db.begin_transaction().await;
+    migrate_database_version(
+        &mut global_dbtx.to_ref_nc(),
+        target_db_version,
+        Some(module_instance_id),
+        kind.clone(),
+    )
+    .await?;
+
+    let disk_version = global_dbtx
+        .get_value(&DatabaseVersionKey(module_instance_id))
+        .await;
+
+    let db_version = if let Some(disk_version) = disk_version {
+        let mut current_db_version = disk_version;
+
+        if current_db_version > target_db_version {
+            return Err(anyhow::anyhow!(format!(
+                "On disk database version for module {kind} was higher than the code database version."
+            )));
+        }
+
+        while current_db_version < target_db_version {
+            let active_states =
+                get_active_states(&mut global_dbtx.to_ref_nc(), module_instance_id).await;
```

Comment:
Call: this can be made more efficient by querying for the states first, then filtering per-module

Reply:
Fixed. Now we read the states at the beginning of the migration process and pass the new states to the next version migration.